seismic/xcorqc/utils.py

```diff
@@ -0,0 +1,200 @@
+from mpi4py import MPI
+import os
+import numpy as np
+from scipy.spatial import cKDTree
+from collections import defaultdict
+from obspy import UTCDateTime, read_inventory, Inventory
+from obspy.geodetics.base import gps2dist_azimuth
+
+# define utility functions
+def rtp2xyz(r, theta, phi):
+    xout = np.zeros((r.shape[0], 3))
+    rst = r * np.sin(theta)
+    xout[:, 0] = rst * np.cos(phi)
+    xout[:, 1] = rst * np.sin(phi)
+    xout[:, 2] = r * np.cos(theta)
+    return xout
+# end func
+
+def xyz2rtp(x, y, z):
+    rout = np.zeros((x.shape[0], 3))
+    tmp1 = x * x + y * y
+    tmp2 = tmp1 + z * z
+    rout[0] = np.sqrt(tmp2)
+    rout[1] = np.arctan2(sqrt(tmp1), z)
+    rout[2] = np.arctan2(y, x)
+    return rout
+# end func
+
+def getStationInventory(master_inventory, inventory_cache, netsta):
+    netstaInv = None
+    if (master_inventory):
+        if (inventory_cache is None): inventory_cache = defaultdict(list)
+        net, sta = netsta.split('.')
+
+        if (isinstance(inventory_cache[netsta], Inventory)):
+            netstaInv = inventory_cache[netsta]
+        else:
+            inv = master_inventory.select(network=net, station=sta)
+            if(len(inv.networks)):
+                inventory_cache[netsta] = inv
+                netstaInv = inv
+            # end if
+        # end if
+    # end if
+
+    return netstaInv, inventory_cache
+# end func
+
+def split_list(lst, npartitions):
+    k, m = divmod(len(lst), npartitions)
+    return [lst[i * k + min(i, m):(i + 1) * k + min(i + 1, m)] for i in range(npartitions)]
+# end func
+
+def drop_bogus_traces(st, sampling_rate_cutoff=1):
+    """
+    Removes spurious traces with suspect sampling rates.
+    :param st: Obspy Stream
+    :param sampling_rate_cutoff: sampling rate threshold
+    :return: Input stream is updated inplace
+    """
+    badTraces = [tr for tr in st if tr.stats.sampling_rate < sampling_rate_cutoff]
+
+    for tr in badTraces: st.remove(tr)
+# end func
+
+def _get_stream_00T(fds, net, sta, cha, start_time, end_time,
+                      baz=None, automerge=False, trace_count_threshold=200,
+                      logger=None, verbose=1):
+
+    stations = fds.get_stations(start_time, end_time, network=net, station=sta)
+
+    stations_nch = [s for s in stations if 'N' == s[3][-1].upper() or '1' == s[3][-1]]  # only N channels
+    stations_ech = [s for s in stations if 'E' == s[3][-1].upper() or '2' == s[3][-1]]  # only E channels
+
+    stt = None
+    if (len(stations_nch) > 0 and len(stations_nch) == len(stations_ech)):
```

Comment:
Consider issuing a warning if `len(stations_nch) > 0 and len(stations_nch) != len(stations_ech)`

Reply:
It's a fairly common scenario and we don't want to clutter the logs with too much information. I issue a warning at a higher level, when data for a time-range couldn't be fetched.
I will add more verbose warnings, activated by a log-level parameter, in a different branch where I will be cleaning up ASDFDataset classes.